### Python Preamble

In [ ]:
# Disulfide Bond Analysis
# Author: Eric G. Suchanek, PhD.
# Last revision: 1/16/23 -egs-
# Cα Cβ Sγ

import math
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt

import plotly_express as px
import seaborn as sns

import proteusPy
from proteusPy import *
from proteusPy.Disulfide import *

from proteusPy.ProteusGlobals import *

import pandas as pd
import pyvista as pv
from pyvista import set_plot_theme

plt.style.use('dark_background')

pv.set_jupyter_backend('ipyvtklink')
set_plot_theme('dark')

# the locations below represent the actual location on the dev drive.
# location for PDB repository
PDB_BASE = '/Users/egs/PDB/'

# location of cleaned PDB files
PDB = '/Users/egs/PDB/good/'

# location of the compressed Disulfide .pkl files
MODELS = f'{PDB_BASE}models/'


## Analysis of Disulfide Bonds in Proteins Within the RCSB Protein Data Bank
*Eric G. Suchanek, PhD. (suchanek@mac.com)* <br> <br>

## Summary
I describe the results of a structural analysis of Disulfide bonds contained in 36,362 proteins within the RCSB Protein databank, https://www.rcsb.org. These protein structures contained 294,478 Disulfide Bonds.  The analysis utilizes Python functions from my ``ProteusPy`` package https://github.com/suchanek/proteusPy/, which is built upon the excellent ``BioPython`` library (https://www.biopython.org). 

This work represents a reprise of my original Disulfide modeling analysis conducted in 1986 ([publications](#publications) item 1) as part of my dissertation. Given the original Disulfide database contained only 2xx Disulfide Bonds I felt it would be interesting to revisit the RCSB and mine the thousands of new structures. The initial results are described in the cells below.

### Requirements
 - Biopython patched version, or my delta applied
 - proteusPy: https://github.com/suchanek/proteusPy/

## Introduction
Disulfide bonds are important covalent stabilizing elements in proteins. They are formed when two Sulphur-containing Cysteine (Cys) amino acid residues are close enough and in the correct geometry to form a S-S covalent bond with their terminal sidechain Sγ atoms. Disulfide bonds most commonly occur between alpha helices and greatly enhance a protein's stability to denaturation. 


## Download PDB Files containing Disulfide Bonds

The RCSB query yielded 

In [ ]:

# Download_Disulfides(pdb_home=PDB_BASE, model_home=MODELS, reset=False)


## Extract the Disulfides from the PDB files
The function ``Extract_Disulfides()`` processes all the .ent files in ``PDB_DIR`` and creates two .pkl files representing the Disulfide bonds contained in the scanned directory. In addition, a .csv file containing problem IDs is written if any are found. The .pkl files are consumed by the ``DisulfideLoader`` class and are considered private. You'll see numerous warnings during the scan. Files that are unparsable are removed and their IDs are logged to the problem_id.csv file. The default file locations are stored in the file globals.py and are the used by ``DisulfideExtractor()`` in the absence of arguments passed. The Disulfide parser is very stringent and will reject disulfide bonds with missing atoms or disordered atoms.


Outputs are saved in ``MODEL_DIR``:
1) ``SS_PICKLE_FILE``: The ``DisulfideList`` of ``Disulfide`` objects initialized from the PDB file scan, needed by the ``DisulfideLoader()`` class.
2) ``SS_DICT_PICKLE_FILE``: the ``Dict Disulfide`` objects also needed by the ``DisulfideLoader()`` class
3) ``PROBLEM_ID_FILE``: a .csv containining the problem ids.

In general, the process only needs to be run once for a full scan. Setting the ``numb`` argument to -1 scans the entire directory. Entering a positive number allows parsing a subset of the dataset, which is useful when debugging. Setting ``verbose`` enables verbose messages. Setting ``quiet`` to ``True`` disables all warnings.

NB: A extraction of the initial disulfide bond-containing files (> 36000 files) takes about 1.25 hours on a 2020 MacbookPro with M1 Pro chip, 16GB RAM, 1TB SSD. The resulting .pkl files consume approximately 1GB of disk space, and equivalent RAM used when loaded.

In [ ]:


#Extract_Disulfides(numb=5000, pdbdir=PDB, datadir=MODELS, verbose=False, quiet=True)


## Load the Disulfide Data
Now that the Disulfides have been extracted and the Disulfide .pkl files have been created we can load them into memory using the DisulfideLoader() class. This class stores the Disulfides internally as a DisulfideList and a dict. Array indexing operations including slicing have been overloaded, enabling straightforward access to the Disulfide bonds, both in aggregate and by residue. After loading the .pkl files the Class creates a Pandas ``DataFrame`` object consisting of the Disulfide ID, all sidechain dihedral angles, the local coordinates for the Disulfide and the computed Disulfide bond torsional energy.

NB: Loading the data takes 2.5 minutes on my MacbookPro. Be patient if it seems to take a long time to load.

In [ ]:
# when running from the repo the local copy of the Disulfides is in ../pdb/models
# PDB_BASE = '../pdb/'

# location of the compressed Disulfide .pkl files
# MODELS = f'{PDB_BASE}models/'

PDB_SS = DisulfideLoader(verbose=True)


The Disulfide and DisulfideList classes include rendering capabiities using the excellent PyVista interface to the VTK package. (http://pyvista.org). The following cell displays the first Disulfide bond in the database in ball-and stick style. Atoms are colored by atom type:
- Grey = Carbon
- Blue = Nitrogen
- Red = Oxygen
- Yellow = Sulfur
- White = Previous residue carbonyl carbon and next residue amino Nitrogen. (more on this below).

The display is actually interactive. Select drag to rotate, mousewheel to zoom. The X-Y-Z widget in the window upper right allows orientation against the X, Y and Z axes.

In [ ]:
ss = PDB_SS[0]
ss.display(style='bs')

The Disulfide and DisufideLoader classes can render Disulfides and DisulfideLists in several molecular styles

In [ ]:
ss.display(style='cpk')

## Examine the Disulfide $C\alpha-C\alpha$ Distances
The Disulfide Bond $Cα-Cα$ distances are constrained due to the overall bond lengths and bond angles of the disulfide bond itself. We can get an overall sense of the protein structure data quality by looking at the distance distribution and removing any disulfides that have unreasonable/physically impossible distances.

In [18]:
#
# split out the left/right handed torsions from the dataframe
#
# retrieve the torsions dataframe
from proteusPy.Disulfide import Torsion_DF_Cols

_SSdf = PDB_SS.getTorsions()

# there are a few structures with bad SSBonds. Their
# CA distances are > 7.0. We remove them from consideration
# below

_far = _SSdf['ca_distance'] >= 10.0
_near = _SSdf['ca_distance'] < 10.0
_zed = _SSdf['ca_distance'] == 0.0

SS_df_zed = _SSdf[_zed].copy()
SS_df_Far = _SSdf[_far].copy()

SS_df = _SSdf.copy()

SS_df = SS_df[Torsion_DF_Cols].copy()
SS_df.describe()


proximal         distal           chi1           chi2  \
count  291274.000000  291274.000000  291274.000000  291274.000000   
mean      234.673833     283.066903     -45.120986      -6.129391   
std       306.557841     307.201849     103.823776     107.659220   
min         1.000000       1.000000    -179.999898    -179.999737   
25%        43.000000      96.000000     -89.180179     -86.688091   
50%       136.000000     194.000000     -64.155480     -54.740993   
75%       318.000000     361.000000     -41.148905      93.320372   
max      8025.000000    8774.000000     179.999904     179.998842   

                chi3           chi4           chi5         energy  \
count  291274.000000  291274.000000  291274.000000  291274.000000   
mean       -4.240517     -24.249724     -30.570362       3.682602   
std        94.106464     110.672137      98.837159       2.382887   
min      -179.981786    -179.999510    -179.999305       0.491737   
25%       -88.157549     -97.617251     -74.409473       2.010495   
50%       -65.393966     -67.597999     -59.775063       3.195969   
75%        93.471721      81.562362      55.534524       4.513430   
max       179.995070     179.999263     179.999693      19.428531   

         ca_distance       phi_prox       psi_prox       phi_dist  \
count  291274.000000  291274.000000  291274.000000  291274.000000   
mean        5.628007    -100.956438      68.687832     -98.326758   
std         1.405404      43.308305      95.624649      42.032985   
min         2.706247    -180.000000    -180.000000    -180.000000   
25%         5.128755    -130.964093     -19.064597    -123.811058   
50%         5.692145    -103.721716     115.442125    -100.394178   
75%         6.233000     -71.638609     141.332085     -73.707005   
max       158.786762     179.970128     179.999324     179.931085   

            psi_dist  
count  291274.000000  
mean       73.146533  
std        94.317661  
min      -180.000000  
25%       -13.958893  
50%       122.484717  
75%       144.709674  
max       179.999470

The average Ca-Ca distance for the entire dataset is 5.62 $\AA$, with a minimum distance of 2.71 $\AA$ and a maximum of 158.8 $\AA$. Since the latter is not physically possible we should examine it further.

In [19]:
SS_df_Far.describe()

proximal       distal        chi1        chi2        chi3  \
count   236.000000   236.000000  236.000000  236.000000  236.000000   
mean    120.775424   234.182203   -9.334648   25.707042    2.090133   
std     209.813931   747.248015  102.301949   93.235923   85.551752   
min       1.000000     1.000000 -179.824424 -179.835259 -179.734548   
25%      14.000000    45.000000  -70.208755  -31.791046  -58.072153   
50%      39.500000    60.000000  -59.984589   15.686457    3.316012   
75%     135.000000   227.000000   68.206634  112.302884   46.491073   
max    1498.000000  8774.000000  179.681323  179.033585  177.542951   

             chi4        chi5      energy  ca_distance    phi_prox  \
count  236.000000  236.000000  236.000000   236.000000  236.000000   
mean     0.326429  -30.758224    7.306442    38.930139 -100.851424   
std    107.579418  103.407088    2.596245    23.298512   47.877949   
min   -176.487239 -179.807853    1.426076    10.386773 -180.000000   
25%    -77.638297  -75.348379    5.122217    22.024159 -132.670131   
50%    -30.558585  -59.188608    7.402771    31.144402  -81.631154   
75%    116.130895   37.455621    8.946429    51.509222  -65.509807   
max    169.032022  179.742066   14.761692   158.786762  170.612433   

         psi_prox    phi_dist    psi_dist  
count  236.000000  236.000000  236.000000  
mean    35.338806 -104.382445   26.408758  
std    123.325830   44.991529  122.516416  
min   -180.000000 -180.000000 -180.000000  
25%    -43.319459 -130.041687  -45.521485  
50%    106.399861 -100.842918   33.657848  
75%    140.995512  -64.583148  129.410256  
max    171.910637   84.092757  179.681140

In [15]:
# The distances are held in the overall Torsions array. We get this and sort
distances = PDB_SS.getTorsions()
distances.sort_values(by=['ca_distance'], ascending=False, inplace=True)

distances.head(10)

source             ss_id  proximal  distal        chi1        chi2  \
93084    2v0o    2v0o_147C_273C       147     273  -99.016165  136.902602   
272024   4bog    4bog_142A_130M       142     130 -115.136243  -49.075789   
113169   6qfo  6qfo_1064A_1685A      1064    1685 -179.824424   -4.931770   
101090   7a3t    7a3t_107A_311A       107     311  -96.864908 -104.377781   
32730    2aye    2aye_298C_295F       298     295  172.773199  -60.214009   
32729    2aye    2aye_295C_298F       295     298  -61.124378  159.401335   
17134    4nzd       4nzd_90A_1C        90       1  -37.429263  131.229281   
236449   4b9d    4b9d_276A_163B       276     163  -51.941249   18.249202   
240299   5jhq     5jhq_466A_12I       466      12  -90.140963  -23.957801   
242547   4apc    4apc_276A_163B       276     163  -59.025935   22.458622   

              chi3        chi4        chi5     energy  ca_distance  \
93084  -124.480681 -101.977868   49.156824   9.870455   158.786762   
272024  107.325187   47.767197 -126.294399   9.878658   127.766571   
113169  109.293963  103.132493 -158.962482   6.569537   101.701154   
101090  -66.261420 -125.732603  -93.089547   9.830990    96.975674   
32730   -69.245084  153.071899  -69.947059   2.194566    93.307879   
32729   -58.878269  -51.122235  166.007732   3.021158    93.102007   
17134  -119.919428   18.145501  -61.886870   7.606388    86.762092   
236449   14.227833   26.655304  -58.784930  10.549097    82.045527   
240299   61.744270  -33.946144  -62.571170   5.710128    81.596937   
242547   27.279432   10.061586  -34.276654  11.022289    81.524647   

          phi_prox    psi_prox    phi_dist    psi_dist  
93084   -67.512488  -44.768308  -69.787164  154.479520  
272024  -49.225505  104.363539  -72.166985  140.944942  
113169  -56.230415  141.426088 -165.267183  116.994438  
101090 -127.716043 -168.280880 -127.290605  130.095138  
32730   -61.589604  -48.450828  -59.163995  -44.851135  
32729   -62.644845  -35.573643  -61.330990  -46.311526  
17134  -180.000000 -180.000000 -180.000000 -180.000000  
236449 -124.249141  -46.418951  -54.081154  -41.609866  
240299  -56.467908  -39.302097  -78.584228  147.584680  
242547 -120.985399  -43.234487  -56.458976  -45.710390

In [17]:
longest = PDB_SS.get_by_name('2v0o_147C_273C')
longest.display(style='sb')

ViewInteractiveWidget(height=1024, layout=Layout(height='auto', width='100%'), width=1024)

## Examining Disulfide Energies


In [ ]:
labels = {'value': 'Energy', 'variable': 'Count'}
cols = ['energy']
px.histogram(SS_df['energy'], labels=labels, histnorm='probability')


### Find the lowest and highest energy disulfides and display them

In [ ]:
All_SS_list = DisulfideList([], 'tmp')
All_SS_list = PDB_SS.getlist()

ssMin, ssMax = All_SS_list.minmax()


In [ ]:
ssMax.display(style='bs')


In [ ]:
ssMin.display(style='bs')

In [ ]:
duolist = DisulfideList([ssMin, ssMax], 'duo')
duolist.display(style='sb')

The structure above shows that the previous and next residues have missing atoms (the long vectors drawn).

In [ ]:
SS_df = SS_df.sort_values(by=['energy'])
#print(f'{SS_df.head(20)}')
badlist = SS_df['ss_id']
print(f'{badlist[20:]}')

bad_SS_list = DisulfideList([],'20 top high energy')
ss = Disulfide()

for i in range(20):
    ssid = badlist[i]
    ss = PDB_SS.get_by_name(ssid)
    bad_SS_list.append(ss)
    #print(f'Energy: {ss.energy}')

#bad_SS_list


In [ ]:
labels = {'value': 'Ca Distance', 'variable': 'Count'}
cols = ['ca_distance']
data = SS_df[cols].copy()
px.histogram(SS_df['ca_distance'], labels=labels)
#px.histogram(SS_df[cols], labels=labels)


## Examine the Disulfide Torsions
The disulfide bond's overall conformation is defined by the sidechain dihedral angles $\chi_{1}$-$\chi_{5}$. Since the S-S bond has electron delocalization, it exhibits some double-bond character with strong minima at $+90°$ and $-90°$. The *Left-handed* Disulfides have $\chi_{3}$ < 0.0° and the *Right-handed* have a $\chi_{3}$ > 0.0°.

These torsion values along with the approximate torsional energy are stored in the DisulfideLoader() class and individually within each Disulfide object. We access them via the ``DisulfideList.getTorsions()`` function.

We can get a quick look at their overall statistics using the ``Pandas.describe()`` function.



In [ ]:
labels = {'value': 'Angle', 'variable': 'Torsion'}
cols = ['chi1', 'chi5']
data = SS_df[cols].copy()
data.set_index('chi1')

px.histogram(SS_df, x='chi1', y='chi5', labels=labels, histfunc='avg', nbins=90)
px.histogram(SS_df[cols], labels=labels)


### Examining torsions by Disulfide Handedness
We split the dataset into these two families easily with Pandas.

In [ ]:
# make two dataframes containing left handed and right handed disulfides

_left = SS_df['chi3'] <= 0.0
_right = SS_df['chi3'] > 0.0

# left handed and right handed torsion dataframes
SS_df_Left = SS_df[_left]
SS_df_Right = SS_df[_right]


In [ ]:
# routine creates 2 lists  for left-handed and right-handed disulfides 
ss_list = PDB_SS.getlist()
left_handed_SS = DisulfideList([], 'left_handed')
right_handed_SS = DisulfideList([], 'right_handed')

i = 0

for i in range(0, len(ss_list)):
    ss = ss_list[i]
    if ss.chi3 < 0:
        left_handed_SS.append(ss)
    else:
        right_handed_SS.append(ss)


print(f'Left Handed: {len(left_handed_SS)}, Right Handed: {len(right_handed_SS)}')



In [ ]:
px.histogram(SS_df, x='chi2', y='chi4', labels=labels, histfunc='avg', nbins=90)

In [ ]:
labels2 = {'value': 'Angle', 'variable': 'Distance'}
px.histogram(SS_df, x='ca_distance', labels=labels2, nbins=100)

We can use a hexbin plot to generate the equivalent of a heatmap to explore the symmetry for the Chi1-Chi5 and Chi2-Chi4 dihedral angles. Since the disulfide bond is symmetric about Chi3 (the S-S bond), one would expect the distribution of Chi1 to be similar to Chi5, and the same with Chi2 and Chi4.

In [ ]:

# Set the figure sizes and axis limits.
DPI = 220
WIDTH = 6.0
HEIGHT = 3.0
TORMIN = -179.0
TORMAX = 180.0
GRIDSIZE = 20


In [ ]:
x = SS_df['ca_distance']
plt.rcParams.update({'font.size': 8})
plt.rcParams['text.usetex'] = True

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)
fig.suptitle(r'$\chi_{1} - \chi_{5}$ Symmetry', fontsize=8)

ax1= fig.subplots()
ax1.set_xlabel(r'Distance')
ax1.set_ylabel(r'Frequency')

ax1.hist(x)

plt.show()


In [ ]:
# !!!

x = SS_df_Left['chi1']
y = SS_df_Left['chi5']

x2 = SS_df_Right['chi1']
y2 = SS_df_Right['chi5']

plt.rcParams.update({'font.size': 8})
plt.rcParams['text.usetex'] = True

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)

ax1, ax2 = fig.subplots(1, 2, sharey=True)

fig.suptitle(r'$\chi_{1} - \chi_{5}$ Symmetry', fontsize=8)

ax1.tick_params(axis='both', which='major', labelsize=6)
ax2.tick_params(axis='both', which='major', labelsize=6)

ax1.set_xlim(TORMIN-1, TORMAX+1)
ax1.set_ylim(TORMIN-1, TORMAX+1)

ax1.xaxis.set_ticks(numpy.arange(TORMIN-1, TORMAX+1, 60))

ax1.set_xlabel(r'$\chi_{1 (left-handed)}$')
ax1.set_ylabel(r'$\chi_{5}$')

ax1.hexbin(x, y, gridsize=GRIDSIZE, cmap='nipy_spectral')

ax2.set_xlim(TORMIN, TORMAX)
ax2.set_ylim(TORMIN-1, TORMAX+1)

ax2.xaxis.set_ticks(numpy.arange(TORMIN-1, TORMAX+1, 60))

ax2.set_xlabel(r'$\chi_{1 (right-handed)}$')

ax2.hexbin(x2, y2, gridsize=GRIDSIZE, cmap='nipy_spectral')

plt.show()



The distributions look extremely similar for both the left-handed and right-handed populations. Let's look at Chi2 and Chi4 to see if this holds true.

In [ ]:
x = SS_df_Left['chi2']
y = SS_df_Left['chi4']

x2 = SS_df_Right['chi2']
y2 = SS_df_Right['chi4']

plt.rcParams.update({'font.size': 8})
plt.rcParams['text.usetex'] = True

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)

fig.suptitle(r'$\chi_{2} - \chi_{4}$ Symmetry', fontsize=8)

ax1, ax2 = fig.subplots(1, 2, sharey=False)

ax1.tick_params(axis='both', which='major', labelsize=6)
ax2.tick_params(axis='both', which='major', labelsize=6)

ax1.set_xlim(TORMIN-1, TORMAX+1)
ax1.set_ylim(TORMIN-1, TORMAX+1)
ax1.xaxis.set_ticks(numpy.arange(TORMIN-1, TORMAX+1, 60))

ax1.set_xlabel(r'$\chi_{2 (left-handed)}$')
ax1.set_ylabel(r'$\chi_{4}$')

ax1.hexbin(x, y, gridsize=GRIDSIZE, cmap='nipy_spectral')

ax2.set_xlim(TORMIN, TORMAX)
ax2.set_ylim(-180, 180)
ax2.xaxis.set_ticks(numpy.arange(TORMIN-1, TORMAX+1, 60))

ax2.set_xlabel(r'$\chi_{2 (right-handed)}$')
ax2.hexbin(x2, y2, gridsize=GRIDSIZE, cmap='nipy_spectral')

plt.show()



The distributions for left/right show distinct differences. Both show the predicted minimum around -75 degrees, but the right-handed group shows a population of disfuldies in the +90 degree range.

In [ ]:
#

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)
ax1 = plt.subplot(111)

x = SS_df['chi3']
y = SS_df['energy']

ymax = 10.0
plt.hexbin(x, y, gridsize=40, cmap='nipy_spectral')
ax1.set_xlim(TORMIN-1, TORMAX+1)
ax1.set_ylim(y.min(), ymax)
ax1.set_xlabel(r'$\chi_{3}$')
ax1.set_ylabel(r'Energy kcal/mol')
ax1.set_title(r'Energy vs $\chi_{3}$')
ax1.xaxis.set_ticks(numpy.arange(TORMIN-1, TORMAX+1, 30))

plt.show()


In [ ]:
#
TORMAX = 180 # for left handed (Chi3 < 0)
fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)
ax1 = plt.subplot(111)

x = SS_df['chi3']
y = SS_df['ca_distance']

ymax = 8.0
plt.hexbin(x, y, gridsize=80, cmap='nipy_spectral')
ax1.set_xlim(TORMIN-1, TORMAX)
ax1.set_ylim(y.min(), y.max())
ax1.set_xlabel(r'$\chi_{3}$')
ax1.set_ylabel(r'C${\alpha}$ Distance ($\AA$)')
ax1.set_title(r'C${\alpha}$ Distance vs $\chi_{3}$')
ax1.xaxis.set_ticks(numpy.arange(TORMIN-1, TORMAX+1, 30))

plt.show()


In [ ]:
from sklearn.mixture import GaussianMixture
n_clusters = 8

tor_df = SS_df[['chi1', 'chi2', 'chi3', 'chi4', 'chi5']].copy()
tor_df.head(1)
gmm_model = GaussianMixture(n_components=n_clusters)
gmm_model.fit(tor_df)
cluster_labels = gmm_model.predict(tor_df)
X = pd.DataFrame(tor_df)
X['cluster'] = cluster_labels
for k in range(n_clusters):
    data = X[X['cluster'] == k]
    plt.scatter(data['chi1'], data['chi5'], s=2)

plt.show()



In [ ]:
# takes over an hour for full dataset
from sklearn.cluster import SpectralClustering
n_clusters = 8

tor_df = SS_df[['chi1', 'chi2', 'chi3', 'chi4', 'chi5']].copy()
X = tor_df.copy()

scm_model = SpectralClustering(n_clusters=n_clusters, random_state=25,
                                n_neighbors=8, affinity='nearest_neighbors')
# takes 51 min with full dataset
X['cluster'] = scm_model.fit_predict(X[['chi1', 'chi5']])

fig, ax = plt.subplots()
ax.set(title='Spectral Clustering')
sns.scatterplot(x='chi1', y='chi4', data=X, hue='cluster', ax=ax, size=2)




In [ ]:
# takes over an hour for full dataset
from sklearn.cluster import AffinityPropagation
n_clusters = 8

tor_df = SS_df[['chi1', 'chi2', 'chi3', 'chi4', 'chi5']].copy()
X = tor_df.copy()

aff_model = AffinityPropagation(max_iter=600, random_state=25)
# takes 51 min with full dataset
X['cluster'] = aff_model.fit_predict(X[['chi1', 'chi5']])

fig, ax = plt.subplots()
ax.set(title='Affinity Propagation')
sns.scatterplot(x='chi1', y='chi5', data=X, hue='cluster', ax=ax, size=2)


In [ ]:
# takes over an hour for full dataset
from sklearn.cluster import AgglomerativeClustering
n_clusters = 4

tor_df = SS_df[['chi1', 'chi2', 'chi3', 'chi4', 'chi5']].copy()
X = tor_df.copy()

agg_model = AgglomerativeClustering(n_clusters=n_clusters)

X['cluster'] = agg_model.fit_predict(X[['chi1', 'chi5']])

fig, ax = plt.subplots()
ax.set(title='Agglomerative Clustering')
sns.scatterplot(x='chi1', y='chi5', data=X, hue='cluster', ax=ax, size=2)


In [ ]:
labels = {'value': 'Chi3', 'variable': 'Variable'}

x = SS_df['chi3']
y = SS_df['energy']
df = SS_df[['chi3', 'energy']].copy()

energy_hist = numpy.histogram2d(x, y=y, bins=360)
energy_hist = energy_hist

fig = px.histogram(df, labels=labels)

fig.show()

In [ ]:
cols = ['chi1', 'chi5']
px.histogram(SS_df[cols], labels=labels)

In [ ]:
cols = ['chi2', 'chi4']
px.histogram(SS_df[cols], labels=labels)

In [ ]:
px.histogram(SS_df['chi3'], labels=labels)

## Summary
Conformational analysis of 294,222 disulfide bonds in 36,362 proteins contained in the RCSB confirms the predominant conformational classes first described in my initial analysis:
- Left-Handed Spiral
- Right-Handed Hook
- Left-Handed Spiral
  

## Publications
* https://doi.org/10.1021/bi00368a023
* https://doi.org/10.1021/bi00368a024
* https://doi.org/10.1016/0092-8674(92)90140-8
* http://dx.doi.org/10.2174/092986708783330566

## Distribution of Backbone Dihedral Angles $\phi - \psi$

A protein's backbone conformation is defined by the dihdedral angles $\phi$ and $\psi$. The ```proteusPy.Disulfide.ExtractDisulfides()``` function computes these at time of Disulfide extraction.

In [ ]:
GRIDSIZE=40

x = SS_df['phi_prox']
y = SS_df['psi_prox']

x2 = SS_df['phi_dist']
y2 = SS_df['psi_dist']

plt.rcParams.update({'font.size': 8})
plt.rcParams['text.usetex'] = True

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)

fig.suptitle(r'RCSB Disulfide Bond $\phi - \psi$ Dihedral Angles', fontsize=8)

ax1, ax2 = fig.subplots(1, 2, sharey=False)

ax1.tick_params(axis='both', which='major', labelsize=6)
ax2.tick_params(axis='both', which='major', labelsize=6)

ax1.set_xlim(TORMIN-1, TORMAX+1)
ax1.set_ylim(TORMIN-1, TORMAX+1)
ax1.xaxis.set_ticks(numpy.arange(TORMIN-1, TORMAX+1, 60))

ax1.set_xlabel(r'$\phi$ Proximal')
ax1.set_ylabel(r'$\psi$')

ax1.hexbin(x, y, gridsize=GRIDSIZE, cmap='nipy_spectral')

ax2.set_xlim(TORMIN, TORMAX)
ax2.set_ylim(-180, 180)
ax2.xaxis.set_ticks(numpy.arange(TORMIN-1, TORMAX+1, 60))

ax2.set_xlabel(r'$\phi$ Distal')
ax2.hexbin(x2, y2, gridsize=GRIDSIZE, cmap='nipy_spectral')

plt.show()

